File: dart/hashmap/find_least_num_of_unique_ints.dart
class Solution {
  int findLeastNumOfUniqueInts(List<int> arr, int k) {
    var hashMap = <int, int>{};
    for (var num in arr) {
      hashMap[num] = (hashMap[num] ?? 0) + 1;
    }
    var valuesFrequency = hashMap.values.toList();
    valuesFrequency.sort();
    print(hashMap);

    var length = hashMap.length;
    for (var element in valuesFrequency) {
      if (element <= k) {
        length--;
        k -= element;
      } else {
        break;
      }
    }
    return length;
  }
}

void main() {
  var arr = [1, 2, 3, 4, 5, 1, 2, 3];
  var k = 2;
  var obj = Solution();
  var result = obj.findLeastNumOfUniqueInts(arr, k);
  print(result);
}

File: go/hashmap/find_least_num_of_unique_ints.go
package hashmap

import (
	"fmt"
	"sort"
)

func findLeastNumOfUniqueInts(arr []int, k int) int {
	hashMap := make(map[int]int)
	for _, num := range arr {
		hashMap[num]++
	}

	var valuesFrequency []int
	for _, v := range hashMap {
		valuesFrequency = append(valuesFrequency, v)
	}
	sort.Ints(valuesFrequency)

	length := len(hashMap)
	for _, element := range valuesFrequency {
		if element <= k {
			length--
			k -= element
		} else {
			break
		}
	}
	return length
}


func main() {
	arr := []int{5, 5, 4}
	k := 3
	result := findLeastNumOfUniqueInts(arr, k)
	fmt.Println(result)
}
